src/rustup-cli/rustup_mode.rs

```diff
@@ -1038,3 +1098,52 @@ fn set_default_host_triple(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
     cfg.set_default_host_triple(m.value_of("host_triple").expect(""))?;
     Ok(())
 }
+
+#[derive(Copy, Clone, Debug)]
+enum CompletionCommand {
+    Rustup,
+    Cargo,
+}
+
+impl CompletionCommand {
+    fn variants() -> [&'static str; 2] {
+        ["rustup", "cargo"]
+    }
+}
+
+impl FromStr for CompletionCommand {
+    type Err = String;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        match s {
+            _ if s.eq_ignore_ascii_case("rustup") => Ok(CompletionCommand::Rustup),
+            _ if s.eq_ignore_ascii_case("cargo") => Ok(CompletionCommand::Cargo),
+            _ => Err(String::from("[valid values: rustup, cargo]")),
```

Comment:
I dislike that `variants()` enumerates the possible strings, *this* `match` enumerates the strings, **and** this error string enumerates the strings, which match the completion commands.

I'd almost prefer:

```rust
static COMPLETIONS: &[(&'static str, CompletionCommand)] = &[
    ("rustup", CompletionCommand::Rustup),
    ("cargo", CompletionCommand::Cargo),
];
```

combined with a `from_str` which uses that array, and a `variants()` which maps it into a list of strings, and then this error can be constructed from that list too.  All in all, it just feels a little too denormalised for my liking.